### DISCUSSION TITLE
Get path for a sheet

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#1692] [Q&A] [answered]
I'm in a `CsvSheet` or a `DirSheet`. How can I get the full file / directory path to it? (I was told `gv` before but I'm not seeing any column with the path.

**Comment by saulpw** [comment, marked as answer]
Try `=row.resolve()` (each row is a Python Path-like object).

**Reply by cool-RR** [reply]
This works in a `DirSheet`, though it only tells me the path of the files, not of the `DirSheet`. I could deduce it from the file paths. But I want a solution that works for any kind of sheet. Is there any reason for sheets not to have an attribute for their source file path? I'd imagine this data is generally useful.

**Reply by saulpw** [reply]
The attribute is `source`, and for a source sheet, you can use `sheet.source.resolve()`.  You can find all sheet attributes via `g Ctrl+Y` which pushes the sheet object itself as a sheet.

**Reply by cool-RR** [reply]
Ah, so I could concievably define a command `copy-path-to-clipboard` that uses the `pathlib` module to join the working directory to `sheet.source`, then the `clipboard` module to copy that file path to the clipboard?

**Reply by cool-RR** [reply]
Ah oops, I don't need `pathlib` because of `resolve`, missed that.

**Reply by saulpw** [reply]
Yes, though `sheet.source.resolve()` should get you an absolute path in the first place, without needing to join the working dir to it.

**Reply by cool-RR** [reply]
I want to bind it to `Ins-f`, i.e. press `Insert` then `f`. This didn't work:

```
vd.allPrefixes = ['g', 'z', '^[', 'Ins']
BaseSheet.addCommand('Insf', 'copy-path-to-clipboard', 'copy_path_to_clipboard')
```

Any idea why?

**Reply by saulpw** [reply]
It's looking for `KEY_IC`, the 'raw' keycode for Insert.  I guess it looks for prefixes before doing the translation.  `Ins` is an interesting prefix character to choose, too.

**Reply by saulpw** [reply]
This is changed in `develop` now (add `Ins` instead of `KEY_IC` to `vd.allPrefixes`); we're ultimately going to change to prettykeys for keybindings also.

**Reply by cool-RR** [reply]
Works! Thank you. I'll wait for a release to be made until I start using it regularly.

I do think it's a bit weird that I have to do this: 

```
Sheet.addCommand('Insp', 'python-shell', 'vd.python_shell()')
```

I mean, it's confusing that `'Insp'` means "Insert then p", but I can live with that.

**Reply by saulpw** [reply]
In general, our documentation uses the following keystroke forms:

- `Key1-Key2`: press `Key1`, release `Key1`, then press `Key2` (`Key2` after `Key1`)
- `Key1+Key2`: press `Key1`, then press `Key2` without releasing `Key1` (`Key2` with `Key1`)

So we have `Ctrl+P` vs `Esc-p`.  So maybe this should be `Ins-p`?  We'd have to make the `-` key be something like `Minus` or `Dash`, which is a little weird, but that would also pave the way for `Backtick` and `Tick` which is what we usually write anyway.  Maybe these are aliases and we can use both.

**Reply by cool-RR** [reply]
I tried to do this now:

```
Sheet.addCommand('Ins-p', 'python-shell', 'vd.python_shell()')
```

But it wouldn't recognize when I pressed `Insert` and then `p`. Not sure whether you meant it's implemented already or not.

**Reply by saulpw** [reply]
Not implemented yet.

**Reply by cool-RR** [reply]
Has "insert" support been implemented in 2.11.1? Or is there a tracking issue for this?

**Reply by saulpw** [reply]
The thing that's not been implemented is the `Ins-p` notation, allowing arbitrary prefixes using `-` as a separator.  For a quick hack, we could just remove any `-` or space characters in the middle of a keybinding.

**Reply by cool-RR** [reply]
But that's not working for me either:

```
Sheet.addCommand('Insp', 'python-shell', 'vd.python_shell()')
```
